File: src/java/engine/movegen/bitboards/OneStep.java
package java.engine.movegen.bitboards;

/**
 * Created by melvic on 7/14/18.
 */
public interface OneStep {
    class Masks {
        public static final long NOT_A_FILE = 0xfefefefefefefefeL;
        public static final long NOT_H_FILE = 0x7f7f7f7f7f7f7f7fL;
        public static final long NOT_AB_FILE = 0xfcfcfcfcfcfcfcfcL;
        public static final long NOT_GH_FILE = 0x3f3f3f3f3f3f3f3fL;
    }

    public long east(long bitboard);
    public long west(long bitboard);
    public long north(long bitboard);
    public long south(long bitboard);
    public long northEast(long bitboard);
    public long northWest(long bitboard);
    public long southEast(long bitboard);
    public long southWest(long bitboard);
}

File: src/engine/movegen/OneStep.scala
package engine.movegen

import java.engine.board.Board

import engine.board.bitboards.Bitboard.U64
import engine.utils.Masks

/**
  * Created by melvic on 8/5/18.
  */
trait OneStep {
  type Step = U64 => U64

  def north: Step
  def south: Step
  def east: Step
  def west: Step

  def northEast: Step = north andThen east
  def northWest: Step = north andThen west
  def southEast: Step = south andThen east
  def southWest: Step = south andThen west
}

trait PostShiftOneStep extends OneStep {
  def north = bitboard => bitboard << Board.SIZE
  def south = bitboard => bitboard >>> Board.SIZE
  def east = bitboard => (bitboard << 1) & Masks.NotAFile
  def west = bitboard => (bitboard >>> 1) & Masks.NotHFile
}